metecho/api/models.py

```diff
@@ -915,6 +915,26 @@ class Task(
     assigned_dev = models.CharField(max_length=50, null=True, blank=True)
     assigned_qa = models.CharField(max_length=50, null=True, blank=True)
 
+    currently_refreshing_dataset_schema = models.BooleanField(default=False)
+    dataset_schema = models.JSONField(
+        default=dict,
+        encoder=DjangoJSONEncoder,
+        blank=True,
+        help_text=_("Cache of the ScratchOrg schema related to this Task"),
+    )
+    currently_refreshing_datasets = models.BooleanField(default=False)
+    datasets = models.JSONField(
+        default=dict,
+        encoder=DjangoJSONEncoder,
+        blank=True,
+        help_text=_("Cache of the dataset definitions from the current Task branch"),
+    )
+    datasets_parse_errors = models.JSONField(
+        blank=True,
+        default=list,
+        help_text=_("User-facing errors that occurred during dataset refresh"),
+    )
```

Comment:
Added this extra field so the frontend can render better errors to the user, instead of a generic "some definitions could not be parsed". The actual errors are populated in `jobs.refresh_datasets` but in summary we detect 3 problems right now:

- The `datasets/` folder is missing altogether
- Extraneous files are found at the top level of `datasets/` (we expect only folders)
- For folders like `datasets/<name>/` we expect a single `*.extract.yml` file. An error is added if we find zero or more than one

What do you think?

Reply:
Let's skip the test for `Extraneous files are found at the top level of datasets/ (we expect only folders)`

Otherwise, good!

An example extract.yml is [here](https://github.com/SFDO-Tooling/CumulusCI/blob/cc97cb30047935723df47029d7f82e865b4f8bcc/datasets/test.extract.yml
)

And example code for parsing it and "flattening" it to a simple form (one declaration per SObject and a flat list of fields) is [here](https://github.com/SFDO-Tooling/CumulusCI/compare/feature/extract-declarations-July#diff-f02fed8530e01d1bc62219017d66ead426ea011b2dff4a12d6d79908d3d430ccR259).

These are on branches that should be merged over the next several days.

Reply:
I've removed the folders-only check in fdef6de63305155a546d2fde6339027f2b62517e.

My understanding was that definitions were already flat like that as explained in the [docs](https://cumulusci.readthedocs.io/en/stable/data.html#defining-datasets), so I was just converting the YAML into JSON. It looks like further processing is needed though. It that something I should start looking into now, or should I just wait for this features in CCI?